RFD 49: desktop access clipboard

zmb3 opened the pull request:
First draft.

zmb3 committed:
RFD 49: desktop access clipboard

first draft

zmb3 commented:
@klizhentas @xinding33 requested your review here as there are some interesting security/user-experience implications.

ibeckermayer commented on rfd/0049-desktop-clipboard.md:
```diff
@@ -0,0 +1,197 @@
+code. Additionally, even with browser-specific code, some browsers simpy won't
+work, as we need:
+
+- access to the clipboard upon receiving a TDP message, which will be decoupled
```
I don't understand what this bullet point means

zmb3 replied:
In some browsers, we can't use clipboard API to write to the clipboard when the browser receives a "clipboard data" TDP message. This is because the browser only allows access to the clipboard in the context of a click or touch event handler.

ibeckermayer replied:
I see, my fault I just wasn't connecting it to the mention of user gestures above.

Regarding this, the MDN documentation is a bit unclear. For example, we're told of the [Clipboard API](https://developer.mozilla.org/en-US/docs/Web/API/Clipboard)

> Calls to the methods of the Clipboard object will not succeed if the user hasn't granted the needed permissions using the Permissions API and the "clipboard-read" or "clipboard-write" permission as appropriate.

If we look at the [Permissions API Browser compatibility table](https://developer.mozilla.org/en-US/docs/Web/API/Permissions_API#browser_compatibility), it looks like "clipboard-read" and "clipboard-write" are hard no's for Firefox and Safari. However, if we look at the [Clipboard's Browser compatibility table](https://developer.mozilla.org/en-US/docs/Web/API/Clipboard_API#clipboard), it looks like `readText` and `writeText` are fully supported in all browsers except Firefox. I'm not sure how to square that circle with Safari.

Additionally, I only saw a note on Firefox that it's extensions require user-initiated event callbacks for `writeText`'s, however above you indicate that this is also the case for Safari.

ibeckermayer commented on rfd/0049-desktop-clipboard.md:
```diff
@@ -0,0 +1,197 @@
+
+- access to the clipboard upon receiving a TDP message, which will be decoupled
+  from the user gesture
+- to detect copy actions that were performed outside the browser
```
Can't this be solved by polling, or `onmouseenter`?

zmb3 replied:
On Chrome, yes. On Firefox or Safari no, because mouse enter doesn't count as a user gesture.

ibeckermayer replied:
I see. It's somewhat ugly, but one option is to add a "sync clipboards" button.

ibeckermayer commented on rfd/0049-desktop-clipboard.md:
```diff
@@ -0,0 +1,197 @@
+4. File List: a list of files to be transferred
+5. File Stream: the contents of a file, allowing transfer of specific chunks
+
+Initially, Teleport will only support option 1 - generic clipboard data. Options
```
This comment seems to perhaps imply that image copy/pastes will be considered separate from other file-type copy-pastes. Is this the case? Will the file transfer feature not allow image files? Or transfer them as regular binary files, and ignore the Palette/Metafile options?

I think these questions are out of scope for this RFD, but this section may warrant revision given our current understanding of these details.

zmb3 replied:
The file list / file stream type will support any file, images included.

I'm not currently aware of any reason why palette or metafile would be used today.

zmb3 committed:
Update based on latest product decisions

- We will support Chrome only, for now
- We will not implement a separate text box like Guacamole does

zmb3 commented:
@smallinsky can you take a look tomorrow? I'd like to get this merged.

xinding33 commented on rfd/0049-desktop-clipboard.md:
```diff
@@ -0,0 +1,218 @@
+clipboard APIs to implement clipboard sharing in a way that is natural to users
+and won't require them to change their behavior.
+
+Since Chrome has the largest market share of any browser, we will officially
+support clipboard sharing in Chrome. For other browsers, desktop access will
+continue to function as it does today, without clipboard support.
```
Agreed. This decision is consistent with our discussion. Thanks for modifying the RFD!

xinding33 reviewed (approved):
Looks good to me!

zmb3 committed:
Merge branch 'master' into zmb3/rfd-clipboard-sharing
